### PR TITLE
Revise ZEP0

### PR DESCRIPTION
**Pull request by MSanKeys963**
Hi all. 👋🏻 

I'm proposing changes to our ZEP process, i.e. ZEP0, based on the discussions at #44, #52 and #55. Here are the major changes I'm proposing through this PR:

1. Addition of new ZEP type under Specification ZEPs category → Protocol Extension ZEPs
> Until now, we have only been using Core ZEPs to make changes to the specification. Extension ZEPs would help define and clearly add new extensions to the specification via the extension mechanism.
2. Adding a new ZEP category → Lean ZEPs
> This category is suitable for adding significantly small and straightforward changes to the Zarr specification directly via a pull request.
3. The process of submitting a ZEP
> Until now, if anyone wishes to submit a ZEP proposal, they'd need to create two PRs, first in zeps repository and the other in zarr-specs repository. I've proposed simplifying this process by combining the zeps and zarr-specs repositories. Now, the author could submit only one PR to submit a ZEP proposal for the Zarr specification.
4. Discussing a ZEP
> As pointed out in #55, discussing a ZEP proposal in multiple issues and PR fragments the conversation, leading to a loss of motivation and momentum. I've simplified this and proposed that all the discussions occur in a single PR. I've also kept an option to open separate issues if the conversation becomes too long and hard to follow, but this depends on the need and the author's choice. But please keep in mind that all the conversation and issues will happen in zarr-specs repository.
5. How does a ZEP become accepted?
> This is mostly formalising the various phases I've outlined in https://github.com/zarr-developers/zeps/issues/55#issuecomment-1905157744. Until now, we have been using a voting mechanism to decide on the fate of the ZEP proposals. Over the past couple of years, we've realised we need more involved participation from the stakeholders. I've proposed to add a couple of phases before we start voting on the ZEPs.
6. ZEP Format
> I've proposed that the ZEP be in `.rst` instead of `.md` format, similar to the specification text format.

There is also a question regarding the versioning of the ZEP proposal, which are modified after acceptance, for example ZEP0. There's a suggestion of using semantic versioning by @LDeakin in https://github.com/zarr-developers/zeps/issues/55#issuecomment-1962288366, which is something we can discuss and adopt.
PS. If we adopt semantic versioning, this update would be ZEP0.1. :) 

Many thanks to @joshmoore, @rabernat, @jbms and others for the initial discussions regarding the ZEP0 revision.

Also, I've been fiddling around to merge the zeps and zarr-specs repository and have come up with: https://docs-test-sanket.readthedocs.io/en/latest/. I'd appreciate any feedback or comments on the new webpage design.

Please let me know what you think of this proposal. Also, please send this to folks interested in this conversation.

CC: @zarr-developers/implementation-council @zarr-developers/steering-council

### CODE REVIEW

File: active/ZEP0000.md [hunk 1]
```diff
@@ -60,29 +64,53 @@ that require collaboration across multiple projects.
...
+extension ZEPs) are part of Spec ZEPs and apply to Zarr storage extension
+specification and various implementations under the Zarr Implementations. The
+extension ZEPs are designed so that they are not necessary to be adopted by
+every other Zarr implementation. The extension ZEPs may or may not include
+involvement from the author of various Zarr implementations and the general
+Zarr community. The general idea of having an extension ZEP is to add
+functionality, data types and features to accommodate emerging and essential
+use cases.
+
+- **Lean ZEP**
```
Comment (normanrz) [inline active/ZEP0000.md:101]: Can you provide an example for a Lean ZEP? Where would adding a codec fall?

File: active/ZEP0000.md [hunk 2]
```diff
@@ -264,130 +299,217 @@ original and new ZEPs respectively. Process ZEPs may also have a status of
...
-If there are substantive objections, then the ZEP remains in `Draft` state,
-discussion continues as normal, and it can be proposed for acceptance again
-later once the objections are resolved.
+The Reading Phase, or Request for Comments (RFC) phase, initiates the ZEP
+proposal process, marking the introduction of the proposal to the Zarr
+community.
 
-In unusual cases, the [Zarr Steering Council] may be asked to decide whether a
-controversial ZEP is `Accepted`.
+1. **Stakeholder Engagement**:
```
Comment (normanrz) [inline active/ZEP0000.md:314]: A reader might ask: Who are the stakeholders and how do I contact them?

File: active/ZEP0000.md (hunk 1 above)
Comment (rabernat) [reply active/ZEP0000.md:101]: I agree with this comment. In particular, how will an author choose between an Extension ZEP and an Lean ZEP?

File: active/ZEP0000.md [hunk 3]
```diff
@@ -137,32 +164,51 @@ mentioned below. This allows the author to flesh out the draft ZEP to make it
...
-For SPEC ZEPs, the proposal should be submitted as a draft ZEP via two GitHub
-pull requests, one to the [ZEPs] repo and the other to the [zarr-specs] repository.
+### For SPEC ZEPs
+
+The proposal should be submitted as a draft ZEP via a GitHub pull
+request to the [zarr-specs] repository.
+
+The PR should contain the narrative document explaining the need, importance,
+and use case for the change of the ZEP and should be submitted in the
+zarr-specs repository with the name `zep-<n>: <title>.rst` under
```
[the remark this answers is not in the document]
Comment (rabernat) [reply active/ZEP0000.md:176]: I would love to move to markdown.

File: active/ZEP0000.md [hunk 4]
```diff
@@ -137,32 +164,51 @@ mentioned below. This allows the author to flesh out the draft ZEP to make it
...
+
+### For Lean ZEPs
+
+The proposal should be submitted as a draft ZEP via a GitHub pull
+request to the [zarr-specs] repository.
+
+The PR should contain the specific changes to the Zarr specification without
+requiring a separate narrative document. As Lean ZEPs involve minor and
+non-controversial changes, a narrative document and the use of the template and
+instructions file are not necessary.
```
Comment (rabernat) [inline active/ZEP0000.md:197]: This sounds like a great way to do ZEPs. Everything in one place! 🎉

**Review by rabernat — COMMENTED** [review]
Thanks for doing this Sanket! I think the introduction of the "lean ZEP" is a major improvement in process. Having all the changes and discussion in a single-threaded PR will make it much easier to discuss and review ZEPs.

I see no reason to limit the lean ZEP process to only minor changes. If anything, the more significant the proposal, the more important it is to be able to review and discuss clearly. I predict that, if we adopt this, nearly all future ZEPs will be declared as "lean ZEPs" mostly because it is easier for everyone involved. Who will be the arbiter of whether a ZEP is minor enough to qualify for a Lean ZEP? What will happen if someone submits a Lean ZEP which some people feel is scoped to large? Who will enforce turning it into a regular (old-style) ZEP?

So I would like the see the "lean ZEP" process simply become the process for all ZEPs, rather than the more complicated multiple PR / multiple repo process.  I won't insist on that if there are others who think maintaining the original process is important. However, I'm curious who that is. **Please 👍  this comment if you are in favor of simply making the "lean ZEP" process the default for all ZEPs.** Or :-1: this comment if you would like to keep the multiple repo / multiple PR process for some ZEPs.

**Comment by normanrz** [comment]
I agree with @rabernat. Lean ZEPs sound great!

My learning from the last ZEPs was that we need to engage the stakeholders earlier in the process and not only when presenting them with a final vote. Having the feedback earlier increases buy-in and reduces the number of post-vote changes. I don't think the solution to that problem is more phases and more votes in the process. On the contrary, I think we need to make the conversation easier to follow and actively reach out to ZIC and ZSC. That is why the Lean ZEP approach is appealing. It simplifies the process/paperwork and streamlines the conversation.
In any case, the ZEP process only works with the expectation that members of the ZIC and ZSC bring enough time and attention to follow the conversations. 

I still think we need a mechanism of forcing a vote, though. Consensus seems ill-defined in a distributed, volunteer-driven community.

**Comment by normanrz** [comment]
> There is also a question regarding the versioning of the ZEP proposal, which are modified after acceptance, for example ZEP0. There's a suggestion of using semantic versioning by @LDeakin in [#55 (comment)](https://github.com/zarr-developers/zeps/issues/55#issuecomment-1962288366), which is something we can discuss and adopt.
> PS. If we adopt semantic versioning, this update would be ZEP0.1. :)

In the spirit of copying PRs, I would advocate for keeping linear ZEP numbering. The ZEP process is its own artifact that gets created and modified through ZEPs. ZEP 0 introduced the process. This proposed modification should become its own ZEP.

**Comment by jbms** [comment]
Regarding versioning, as I see it, a "ZEP" would just be the additional process taken for reviewing a non-editorial/non-trivial spec change (i.e. PR).  Therefore, there would be no reason to version a "ZEP" except perhaps in its draft state, where you could potentially identify versions by git commit hashes, because once it is accepted it is merged into the spec and any further revisions would be done as a separate PR.  In fact I would propose that the ZEP number be assigned based on the PR number.

For example, let's say we add a "pack_bits" codec in PR 1234, that initially only supports "bool" data type.  Then someone in PR 1255 adds support for a "uint4" data type and modifies the description of the "pack_bits" codec to also support "uint4".  When we document what features are supported by each implementation, if not already explicit we could at this point update the descriptions to indicate that certain implementations support pack_bits with bool only while others support both bool and uint4.

It might be useful to define as part of the spec a precise way to indicate support for various features.  For example we could say that support for a given data type can be indicated by the "codecs/pack_bits/<data_type>" feature identifier.  I'm not sure how important it is to specify these feature flags precisely, though.

**Comment by normanrz** [comment]
I think it is important to separate support for a feature of the Zarr specification in an implemention from the ZEP that introduced the feature.

> It might be useful to define as part of the spec a precise way to indicate support for various features. For example we could say that support for a given data type can be indicated by the "codecs/pack_bits/<data_type>" feature identifier. I'm not sure how important it is to specify these feature flags precisely, though.

I think an implementation needs to be able to decide during metadata parsing whether it can correctly understand a Zarr hierarchy. If it encounters unknown attributes, values or identifiers that would already indicate a lack of support. Basically, that means that new changes can only be additive (e.g. new attributes on existing codecs, new values for attributes, new codecs).

**Comment by jbms** [comment]
> I think it is important to separate support for a feature of the Zarr specification in an implemention from the ZEP that introduced the feature.
> 
> > It might be useful to define as part of the spec a precise way to indicate support for various features. For example we could say that support for a given data type can be indicated by the "codecs/pack_bits/<data_type>" feature identifier. I'm not sure how important it is to specify these feature flags precisely, though.
> 
> I think an implementation needs to be able to decide during metadata parsing whether it can correctly understand a Zarr array/group. If it encounters unknown attributes, values or identifiers that would already indicate a lack of support. Basically, that means that new changes can only be additive (e.g. new attributes on existing codecs, new values for attributes, new codecs).

To be clear, I was imagining that these feature identifiers would be solely for documentation purposes, similar to caniuse.com, not that they would appear in the metadata anywhere.

**Comment by MSanKeys963** [comment]
Thanks all for the feedback so far.

Here are my responses:

> Who will be the arbiter of whether a ZEP is minor enough to qualify for a Lean ZEP? What will happen if someone submits a Lean ZEP which some people feel is scoped too large? Who will enforce turning it into a regular (old-style) ZEP?

@rabernat: The changes which are minor and non-controversial would fall under the Lean ZEP category, e.g. https://github.com/zarr-developers/zarr-specs/pull/256 or adding a new data type. The ZEP, which proposes a significant change to the specification and requires a reference implementation of the sort, should not be submitted under Lean ZEP, e.g. major specification version update (V3 → V4), extensions like sharding, variable chunking, etc.
In the [ZEP workflow](https://zeps--59.org.readthedocs.build/en/59/active/ZEP0000.html#zep-workflow) section, the author has various options to ask around if their changes are minor enough to be considered under a Lean ZEP.

Categorising all ZEPs under Lean might lead to less involved participation in an important proposal from everyone involved. The reading and implementation phases ensure that the stakeholders are given enough time to read and work on implementing the proposal before it is finalised. I've outlined various issues we've encountered here: https://github.com/zarr-developers/zeps/issues/55#issuecomment-1905156595.

I've also proposed removing the option for multiple repo/PR processes so that everything would take place in the zarr-specs repository.

> I don't think the solution to that problem is more phases and more votes in the process.

@normanrz: Why do you think a multi-phase approach would not help have more engaged participation from the stakeholders?

File: active/ZEP0000.md (hunk 1 above)
Comment (MSanKeys963) [reply active/ZEP0000.md:101]: Please see: https://github.com/zarr-developers/zeps/pull/59#issuecomment-1989703623.

File: active/ZEP0000.md (hunk 2 above)
Comment (MSanKeys963) [reply active/ZEP0000.md:314]: Mostly the ZIC and ZSC but also the broader community.
The author is supposed to use the submitted PR to spread the word to everyone, but we also have Google groups set up for notification purposes.

**Comment by normanrz** [comment]
> e.g. [zarr-developers/zarr-specs#256](https://github.com/zarr-developers/zarr-specs/pull/256) or adding a new data type

Adding data types or codecs can be controversial, too. I don't think it is possible to know a priori. That is why a community discussion is needed for every ZEP. PRs are a great tool for that.

> requires a reference implementation of the sort

Shouldn't all ZEPs come with a reference implementation or at least a proof-of-concept?

> > I don't think the solution to that problem is more phases and more votes in the process.
> 
> @normanrz: Why do you think a multi-phase approach would not help have more engaged participation from the stakeholders?

Everybody in the Zarr community is volunteering and very busy. Multi-phase processes require a lot of effort to understand and to follow. I might be wrong but I think that more complex processes can be intimidating and annoying both on the side of the ZEP authors, who need to chase the feedback/votes, and on the side of the community members, who are frequently pressured to issue feedback/votes.

**Comment by MSanKeys963** [comment]
Resurrecting this PR after having discussions and feedback from various community and ZEP meetings. I have extended the scope of Lean ZEPs to add breaking changes if there is unanimous approval from participating stakeholders. This change is specifically targeted to include https://github.com/zarr-developers/zarr-specs/pull/292 and other similar changes in the future.



@normanrz, here are my responses to your last comment —

> Adding data types or codecs can be controversial, too. I don't think it is possible to know a priori. That is why a community discussion is needed for every ZEP. PRs are a great tool for that.

It can be. That's why it's recommended and preferred that an issue in the zarr-specs repo be created to have a community discussion before working on the actual PR for a ZEP; see [here](https://zeps--59.org.readthedocs.build/en/59/active/ZEP0000.html#zep-workflow) - 3rd para.


> Shouldn't all ZEPs come with a reference implementation or at least a proof-of-concept?

ZEPs which add new functionality by changing or extending the existing specification should be accompanied by an implementation/POC.
Other ZEPs like Lean, which includes minor changes (addition or removal of non-controversial changes) or Informational or Process, don't require a POC.


> complex processes can be intimidating and annoying both on the side of the ZEP authors, who need to chase the feedback/votes, and on the side of the community members, who are frequently pressured to issue feedback/votes.

I understand that the phased approach requires involved participation from the active maintainers and implementors. I've outlined issues we've faced with the existing version of the ZEP process at https://github.com/zarr-developers/zeps/issues/55#issuecomment-1905156595 during ZEP1. We need a mechanism to avoid repeating those issues. I think the phased approach is one solution that could help us. If you have suggestions, let's discuss them. I'm all ears. :)

Also, having all the discussions and comments in a single PR for the entirety of a ZEP proposal would be convenient for the author to facilitate feedback and votes from everyone involved.

File: active/ZEP0000.md [hunk 5]
```diff
@@ -60,29 +64,53 @@ that require collaboration across multiple projects.
 
 - **Specification ZEP**
 
-Specification ZEPs deal with the changes related to [Zarr Specifications].
-The SPEC ZEP (abbreviation of Specification ZEP) for now introduces changes in
-the core specification of Zarr. The changes related to core specification follow a
+Specification ZEPs deal with the changes related to [Zarr Specifications]. The
+SPEC ZEP (abbreviation of Specification ZEP) introduces changes to the
```
Comment (d-v-b) [inline active/ZEP0000.md:68]: Unless "SPEC" is an acronym, it should not be all caps.

File: active/ZEP0000.md [hunk 6]
```diff
@@ -60,29 +64,53 @@ that require collaboration across multiple projects.
...
+specification of Zarr. The changes related to specification follow a
 strict and thorough review process and should be adopted by everyone in the
-Zarr community.
+Zarr community. Specification ZEPs are of two types. They are:
 
 - **Core protocol ZEP**
```
Comment (d-v-b) [inline active/ZEP0000.md:73]: what is the difference between a specification zep and a core protocol zep?

Comment (d-v-b) [reply active/ZEP0000.md:73]: ok I see, a core protocol zep is a subtype of a specification zep. What does the word "protocol" mean here? Are we just using that word to avoid re-using the word "specification"?

**Comment by normanrz** [comment]
Thanks for addressing some of our feedback and revising the ZEP.

I think the lean ZEPs help to move the process in the right direction. However, a main problem we are currently facing is lack of (timely) responses from the council members. Without solving that problem, I don't think we will see significant improvements to the process. So, I think it is important to also move https://github.com/zarr-developers/governance/pull/44 forward.

**Comment by d-v-b** [comment]
It seems like a "lean ZEP" as defined in this PR is a spec change that gets approved without a formal ZEP process, and so I would argue that a "lean ZEP" is is not a ZEP, or at least it communicates very little to categorize such changes as ZEPs. 

Instead of requiring that every spec change fall into some ZEP category, It would be simpler to reserve the concept of "ZEP" for formal proposals that do observe a ZEP process. Instead of defining a "lean ZEP", this document can simply state that not all spec changes happen via ZEPs -- some spec changes, like typo fixes, or even semantically significant changes, can be proposed, discussed, and accepted via github PRs provided there is an exceptional level of consensus, bypassing the ZEP process completely.

@normanrz wrote:
> I think the lean ZEPs help to move the process in the right direction. However, a main problem we are currently facing is lack of (timely) responses from the council members. Without solving that problem, I don't think we will see significant improvements to the process. So, I think it is important to also move [zarr-developers/governance#44](https://github.com/zarr-developers/governance/pull/44) forward.

I agree with this completely. We are not bottlenecked by the lack of new ZEP categories (although it's important to define those carefully). We are bottlenecked because there is not enough developer time going into the spec.

**Comment by jhamman** [comment]
> We are not bottlenecked by the lack of new ZEP categories (although it's important to define those carefully). 

Extrapolating on this a bit, let me introduce a new idea here. Rather than adding additional ZEP categories, we do away with all but core-protocol ZEPs, reserving these for major and/or breaking changes to the specification. All other changes (e.g. bug fixes, clarifications) and/or extensions to the specification can be done as pull requests directly to the spec repo and merged through a much simpler process (ideally managed entirely by the spec core team -- not the ZEP process).

**Comment by joshmoore** [comment]
I think the primary additions I'd make to your comment, @jhamman, would be:

- We could probably use a pretty good common definition, or at least understanding, of what a breaking specification change is, including from [provisional versions](https://github.com/zarr-developers/zarr-specs/pull/292/files#diff-39bde03e0d9bdcd6ea478965c1f6ec50785744ce98ec8f3a1d5c56218041f17fR1699). Ultimately, this might come down to a question of versioning.
- For extensions, I think it's important to separate between the implenentation of extensions (adding a codec ✅) and their definition. The latter is designing an interface and I think having wider feedback on those would be import for the health of the ecosystem (cf. https://github.com/zarr-developers/zarr-specs/pull/309#issuecomment-2312757702)

**Comment by MSanKeys963** [comment]
Thanks, everyone, for your feedback. 

I understand the importance of adding changes to the spec via GitHub PRs. I'm in favour, but I'd like to keep the Lean ZEPs category in the process, as it'll help us maintain a record of changes that have gone in the spec in a readable manner — the original idea I proposed in this PR to combine `zarr-specs` and `ZEP` repository and maintain a single website for both of them (sample [here](https://docs-test-sanket.readthedocs.io/en/latest/zeps.html)). The webpage would host and display a list of Lean ZEPs (which would essentially grab text from the PRs, i.e. lean ZEPs). 

Let me know what you think. ^

---

The discussion at https://github.com/zarr-developers/governance/pull/44 is important, and I like what Josh proposed in https://github.com/zarr-developers/governance/pull/44#issuecomment-2352788442. 

I'd like to propose merging this PR unless strong objections exist to this current version. Successive PRs can be added for more changes or to reflect the progress/conclusion from https://github.com/zarr-developers/governance/pull/44.

File: active/ZEP0000.md [hunk 7]
```diff
@@ -60,29 +64,53 @@ that require collaboration across multiple projects.
...
 
 Describes a ZEP which involves changes in the core specification of Zarr. Core
 protocol ZEPs (commonly known as Core ZEPs) are a part of SPEC ZEPs and apply
```
Comment (d-v-b) [inline active/ZEP0000.md:76]:
```suggestion
protocol ZEPs (hereafter referred to as "Core ZEPs") are a type of Spec ZEPs, and apply
```

File: active/ZEP0000.md [hunk 8]
```diff
@@ -60,29 +64,53 @@ that require collaboration across multiple projects.
...
+the Zarr specification. Lean ZEPs are designed for minor modifications that do
+not require an extensive review process or community consensus. These proposals
+aim to streamline the introduction of small improvements or corrections to the
+Zarr specification. Unlike other ZEP types, Lean ZEPs may be introduced through
+a simplified process, typically by submitting a GitHub pull request directly.
+Lean ZEPs focus on non-controversial changes with minimal impact on existing
+implementations, encouraging community input while prioritizing swift
+implementation and agility. In some cases, Lean ZEPs may include breaking
+changes if there is unanimous approval from all participating stakeholders,
+including the ZIC and the broader Zarr community.
```
Comment (d-v-b) [inline active/ZEP0000.md:113]: This section is unclear on what a lean ZEP actually is (in terms of procedure), and what contributors have to in the lean ZEP process. 

- Is a lean ZEP literally any small PR against the zarr specs repo? In this case, what is the utility of describing this as a ZEP? This reasoning should be stated explicitly, because it is not obvious. 
- Is a lean ZEP _more_ than a small PR against the zarr specs repo, e.g. a PR + some summary document on a web page? If so, who writes that summary document? Is it extra work for the person submitting the PR? This should also be stated explicitly in this text.

Comment (d-v-b) [reply active/ZEP0000.md:113]: closing this, because the content I'm looking for is later on in the document :)

File: active/ZEP0000.md [hunk 9]
```diff
@@ -137,32 +164,51 @@ mentioned below. This allows the author to flesh out the draft ZEP to make it
...
+non-controversial changes, a narrative document and the use of the template and
+instructions file are not necessary.
 
-The first PR should contain the narrative text of the ZEP and should be
-submitted in the zep repository with the name `zep-<n>.md` under [/zeps/draft/]
-where `<n>` is an appropriately assigned four-digit number. The draft ZEP must
-use the [ZEP X - Template and Instructions][template.md] file.
+The Lean ZEPs PR should be submitted in the [zarr-specs] repository with the
+name `<zep-<n>: <title>.md` under [/zarr-specs/docs/specs/...], where `<n>` is
+an appropriately assigned four-digit number.
```
Comment (d-v-b) [inline active/ZEP0000.md:201]: I am confused -- if a lean ZEP does not require a narrative document, what exactly is the `<zep - <n>: <title>.md` document? I thought a lean ZEP was just a PR against the spec, in which case shouldn't the procedure be "open a PR against the spec"?

**Comment by d-v-b** [comment]
> I'd like to propose merging this PR unless strong objections exist to this current version.

I think @jhamman's [proposal to slim down the ZEP process](https://github.com/zarr-developers/zeps/pull/59#issuecomment-2310819674) counts for me as a strong objection. We should at least discuss that as a viable alternative to this PR.

@MSanKeys963 if the main goal is to record all spec changes on the zeps web page, then this can be done without introducing a new ZEP category -- we could display ZEPs and PRs together on that site, with a clear indication of which is which.

File: active/ZEP0000.md (hunk 9 above)
Comment (MSanKeys963) [reply active/ZEP0000.md]: Fixed in https://github.com/zarr-developers/zeps/pull/59/commits/81b7ca2c9c554b77c0587ace192f0cbacb255a0b.

Comment (joshmoore) [reply active/ZEP0000.md]: summarizing last week's discussions including @d-v-b, @MSanKeys963 and me:
 * for the three types of changes:
   - ZEPs
   - "lean ZEP"
   - "just a PR"
 * we talked through some specific examples:
   - add a community codec (i.e. extension) -> just a PR, i.e. "patch"
   - move community codec to core codec -> ZEP (ZIC review), i.e. "minor"
   - choosing a default spelling (US vs UK) -> just a PR *but* might want a ZEP to record a decision
   - v4 -> include the world (ZSC, etc.)
 * we did not find a specific case for "lean ZEP" and instead decided it might be a *technical improvement* (template, bot, code, etc.) which helps expedite generating a ZEP from a simple PR.

File: template/template.md [hunk 10]
```diff
@@ -15,13 +15,11 @@ Author: <list of authors’ real name and email addresses>
 
 Status: < Draft | Active | Accepted | Deferred | Rejected | Withdrawn | Final | Superseded >
 
-Type: <Specification | Informational | Process>
+Type: <Specification | Lean | Informational | Process>
```
Comment (joshmoore) [inline template/template.md:18]:
```suggestion
Type: <Specification | Informational | Process>
```

?

File: active/ZEP0000.md [hunk 11]
```diff
@@ -264,130 +276,220 @@ original and new ZEPs respectively. Process ZEPs may also have a status of
...
+[reStructuredTextPrimer] for more information.
 
 ## Header Preamble
 
 ```
 Author: <list of authors’ real names and email addresses>
 Status: < Draft | Active | Accepted | Deferred | Rejected | Withdrawn | Final | Superseded >
-Type: <Specification | Process | Informational>
-Created: <date created on, in dd-mmm-yyyy format>
+Type: <Specification | Lean | Process | Informational>
```
Comment (joshmoore) [inline active/ZEP0000.md:487]:
```suggestion
Type: <Specification | Process | Informational>
```

File: template/template.md (hunk 10 above)
Comment (joshmoore) [inline template/template.md:18]:
```suggestion
Type: <Specification | Informational | Process>
```

File: active/ZEP0000.md [hunk 12]
```diff
@@ -60,29 +66,41 @@ that require collaboration across multiple projects.
...
+implementation of Zarr and, in general, the overall Zarr community. Core ZEPs
+must go through a thorough review process, including involvement, discussion,
+and voting from the author of the proposed ZEP, Zarr Steering Council, author
+(s) of various Zarr implementations (ZIC), and open-source projects/research
+groups using Zarr and the general Zarr community. The general advice is that
+everyone should be made aware of changes introduced by core ZEPs. Core protocol
+ZEPs require community consensus, and developers or users are typically not
+free to ignore them. 
+
+- **Protocol extension ZEP**
```
Comment (joshmoore) [inline active/ZEP0000.md:92]: This is currently under discussion in https://github.com/zarr-developers/zarr-specs/pull/312 -- I can imagine omitting this section if there are any worries and then adding it back in the future.

Some clarity might be needed on if such a ZEP is *creating* an extension or defining a concrete implementation (in which case a ZEP may not be necessary except for a move to "core").

**Review by joshmoore — COMMENTED** [review]
Few initial comments

**Comment by joshmoore** [comment]
@rabernat, @jakirkham, @alimanfoo, @ryan-williams: are there any changes you would like to see here?

File: active/ZEP0000.md [hunk 13]
```diff
@@ -60,29 +66,41 @@ that require collaboration across multiple projects.
...
+and apply to [Zarr Specifications] and its various implementations under the
+[zarr.dev/implementations]. The core protocol ZEPs should be adopted by every
+implementation of Zarr and, in general, the overall Zarr community. Core ZEPs
+must go through a thorough review process, including involvement, discussion,
+and voting from the author of the proposed ZEP, Zarr Steering Council, author
+(s) of various Zarr implementations (ZIC), and open-source projects/research
+groups using Zarr and the general Zarr community. The general advice is that
+everyone should be made aware of changes introduced by core ZEPs. Core protocol
+ZEPs require community consensus, and developers or users are typically not
+free to ignore them.
```
Comment (rabernat) [inline active/ZEP0000.md:90]: Given the fact that we have decided not to version the V3 spec (see https://github.com/zarr-developers/zarr-specs/issues/277), it would be helpful to state explicitly whether any core protocol ZEP (i.e. "changes in the core specification of Zarr") will generate a new major version of the spec.

In my opinion, these should go together. Any change major enough to warrant a core protocol ZEP would also trigger a major version increment. The next core protocol ZEP will produce Zarr V4.

File: active/ZEP0000.md (hunk 6 above)
Comment (rabernat) [reply active/ZEP0000.md:75]: :+1 to eliminating Protocol and sticking with Specification.

File: active/ZEP0000.md (hunk 12 above)
Comment (rabernat) [reply active/ZEP0000.md:92]: I favor removing Protocol Extension ZEP completely in favor of an extension framework that lies outside the ZEP process (as proposed in https://github.com/zarr-developers/zarr-specs/pull/312)

File: active/ZEP0000.md [hunk 14]
```diff
@@ -137,58 +154,60 @@ mentioned below. This allows the author to flesh out the draft ZEP to make it
...
+The PR should contain the narrative document explaining the need, importance,
+and use case for the change of the ZEP and should be submitted in the
+zarr-specs repository with the name `zep-<n>: <title>.md` under
+[/zarr-specs/docs/zeps/draft], where `<n>` is an appropriately assigned
+four-digit number. The draft ZEP must use the [ZEP X - Template and
+Instructions][template.md] file.
+
+The SPEC ZEPs PR should also contain actual changes to the specification under
+[/zarr-specs/docs/specs/...] and mention the assigned four-digit ZEP number
+from the ZEP document.
```
Comment (rabernat) [inline active/ZEP0000.md:173]: 👏👏👏

This fixes my main complaint about the ZEP process! Thanks!!!

File: active/ZEP0000.md [hunk 15]
```diff
@@ -137,58 +154,60 @@ mentioned below. This allows the author to flesh out the draft ZEP to make it
...
 ## Discussing a ZEP
 
-As soon as the draft ZEP is committed to the ZEP repository, the author
-(s) should create a discussion thread for the ZEP to provide a central place to
-discuss and review its contents. The ZEP author(s) may create an issue in
-[ZEPs] repository for informational and process ZEPs and or similarly create an
-issue in [zarr-specs] repository for specifications ZEPs. 
+As soon as the draft ZEP is committed to the [zarr-specs] repository, the author
+(s) should use the submitted PR for the ZEP as a central place to discuss and
+review its contents. The ZEP author(s) may create additional issues in the
```
Comment (rabernat) [inline active/ZEP0000.md:204]: 👍

File: active/ZEP0000.md (hunk 2 above)
Comment (rabernat) [reply active/ZEP0000.md:293]: :+1: to defining stakeholders clearly.

File: active/ZEP0000.md [hunk 16]
```diff
@@ -264,130 +276,220 @@ original and new ZEPs respectively. Process ZEPs may also have a status of
...
+      as a precautionary measure.
+
+Authors may utilize the submitted pull request (PR) for gathering comments,
+feedback, acknowledgment, and facilitating engagement from ZIC members during
+the Reading Phase.
+
+### Implementation Phase
+
+The Implementation Phase encompasses the timeline during which interested
+adopters actively work on integrating the proposal into their open-source
```
Comment (rabernat) [inline active/ZEP0000.md:327]: Why should we limit implementations to open-source ones? ARC-GIS supports Zarr, for example.

```suggestion
adopters actively work on integrating the proposal into their
```

**Review by rabernat — COMMENTED** [review]
This is great progress. Nice work @MSanKeys963!

I am excited by the simplifications this will bring.

My main feedback

- Let's clarify whether Specification ZEPs generate a new spec major version
- Let's eliminate extension ZEPs entirely

---

Meta question: should this change be done via a ZEP? 🤣

File: active/ZEP0000.md (hunk 13 above)
Comment (MSanKeys963) [reply active/ZEP0000.md]: Done. Added via https://github.com/zarr-developers/zeps/pull/59/commits/3ae1fc5833cc95220e87370c308b20619f4ab931.
